### DISCUSSION TITLE
MSI Z170 + TU116 ReBAR Supported

### DISCUSSION THREAD

**Discussion by UnidentifiedTag** [terminatorul/NvStrapsReBar#3] [General]
Motherboard model: MSI Z170A GAMING M7 (MS-7976)
Motherboard chipset: Z170
Graphics card model: GV-N166SGAMING OC-6GD
GPU chipset: TU116
DeviceID (check GPU-Z) GPU PCI VendorID : 10DE:21C4
Secondary DeviceID (check GPU-Z) GPU PCI subsystem IDs : 1458:4014
VRAM size: 6144 MB (6GB)
New BAR1 size (GPU-Z): 8192 MB (8GB)
New BAR1 Total size (nvidia-smi): 8192 MiB
NVIDIA Driver version : 546.17

![Screenshot 2023-12-14 230235](https://github.com/terminatorul/NvStrapsReBar/assets/54625394/ab9c06b9-0ac9-4227-841e-abddceb0b121)
![Screenshot 2023-12-15 124251](https://github.com/terminatorul/NvStrapsReBar/assets/54625394/4fbad9ef-4891-4183-bb9c-52d12c38408c)
![Screenshot 2023-12-15 130044](https://github.com/terminatorul/NvStrapsReBar/assets/54625394/f2a8c241-8a2f-4b33-9fba-61cbb3c4c3d0)

* This is a great achievement @terminatorul as it marks an unsupported motherboard and unsupported NVIDIA 16 series card supporting ReBAR

**Comment by UnidentifiedTag** [comment]
i've gained an extra 131 points in TIME SPY 🥹which is more than you would ever get in overclocking 🥲 i've also increased my PCI EXPRESS by setting 0x0000000080000000 in nvidiaProfileInspector rBAR - Size Limit! on the _GLOBAL_DRIVER_PROFILE (Base Profile) profile

**Reply by UnidentifiedTag** [reply]
![Screenshot 2023-12-15 152440](https://github.com/terminatorul/NvStrapsReBar/assets/54625394/075dbaf7-5570-4f8d-bfc2-d33fb2a3f89d)

with my OC the score is so high, i would class time spy as player-able hitting 29fps and no lower but above
https://www.3dmark.com/spy/43876766

**Reply by UnidentifiedTag** [reply]
my whole system in timespy under i7 6700k with a 1660 super is ranked so high now, im 9th overall in the leader board

![Screenshot 2023-12-15 153131](https://github.com/terminatorul/NvStrapsReBar/assets/54625394/fde8796c-a836-4543-a4f1-920dbeba6959)

**Comment by UnidentifiedTag** [comment]
Video has had lots of views now, over 3000, I see over here issues are flying in 😂

let me know when you have an update that is worth making a new video about, so we can cut down the time it takes to build the rebar module and make the process more automated. More automated would hopefully mean less issues 👍🏻

**Reply by xCuri0** [reply]
@UnidentifiedTag it should be updated because the current version doesn't require any compiling. just flash modified bios and have the NvStrapsRebar.exe app configure it automatically for you

seems like some people are still trying to compile it and messing up

**Reply by UnidentifiedTag** [reply]
Wow really ill have to look at the documentation

**Reply by UnidentifiedTag** [reply]
So let me get this straight, with 0.3v you just need to download the executable and the module, add the module to the bios file and flash, then use the executable to do the rest?

**Reply by UnidentifiedTag** [reply]
I can’t believe you guys have nailed this in less than a month, this is insane progress from what people had to do with compiling 😦

**Reply by UnidentifiedTag** [reply]
I’ve updated the video description telling everyone to view the GitHub page for now, as the video is now out of date. 

I’ll follow up with a new video some time this month showing the new way and explaining the improvements you guys have made

well done 👏

**Reply by xCuri0** [reply]
@UnidentifiedTag i don't think there's any reason to use the old version even if you want to compile it though.

old version has a bug where resume from sleep doesn't work. 0.3 fixes that

**Reply by UnidentifiedTag** [reply]
You know how some people are, they like to review all code going in to there computer before they apply it. 

So I believe the compile option should stay but definitely the pre built option should be the route almost all users should take

**Reply by terminatorul** [reply]
@UnidentifiedTag 
> I’ve updated the video description telling everyone to view the GitHub page for now, as the video is now out of date.
> 
> I’ll follow up with a new video some time this month showing the new way and explaining the improvements you guys have made
> 
> well done 👏

Hi ! If you want, this time maybe you could write a script before you start shooting new video footage, so then you can let some technical person check it for accuracy before you put in any recording time.

**Reply by UnidentifiedTag** [reply]
Sure if you wish to help with the script you’re input would be welcome 

1st I was thinking about starting the video by showing users how to do the operation

1) download UEFITool non-NE: https://github.com/LongSoft/UEFITool/releases/tag/0.28.0 and use msinfo32 to get the motherboard information to look up the manufacturer for bios, download the bios file from that manufacturer 
2) download 0.3v https://github.com/terminatorul/NvStrapsReBar/releases/download/v0.3/NvStrapsReBar.exe and https://github.com/terminatorul/NvStrapsReBar/releases/download/v0.3/NvStrapsReBar.ffs
3) use UEFITool non-NE to look for GUID `3C1DE39F-D207-408A-AACC-731CFB7F1DD7`
4) go to the last of the dxe drivers and insert the NvStrapsReBar.ffs we downloaded earlier, by insert after method 
5) save the modified bios file put it on a fat32 usb and explain the different ways of flashing “now that we have additional time
6) flash it (maybe show that being done as well
7) talk about the options on the GitHub description when using NvStrapsReBar.exe if you would expand on that and share more information on these options and which are good for certain users I will pass that on in the video
8) apply an option with NvStrapsReBar.exe
9) explain that there is also a compile option (users have been asking me if the same steps in the video for v0.1 can be used for v0.3) could you elaborate on that? I’ll obviously also mention the improvements that have been made and that most people should use this new option in this video instead of the compile option, but if they wish to be tinfoil hats they are welcome to follow the one hour long video

I’d like to add a song/music track, to the video, needs to be copyright free or royalty free with something like a mention 

video will be on hold for a week or two, as it isn’t my computer and I will need to use a capture usb device with a laptop to get the recording. 

Unrelated : I’ve been busy watching the spacex starship launch and I’m currently with out my main rig as I’ve had to sent off my 4090 back to the manufacturer as it has a defect with the Nvidia core, personally I’m fighting for it to be replaced with a new one as it’s a fault that could of easily been spotted if they had of tested the core, you’ll find video clips of that on my YouTube channel and I’ll probably be making a video about that as well to warn others… not an easy fault to spot

**Reply by terminatorul** [reply]
Your plan looks solid !

```
        (E) Enable auto-settings BAR size for known Turing GPUs (GTX 1600 / RTX 2000 line)
```
This option uses a fixed (hard-coded) list of PCI IDs to identify all aprox130 known Turring GPUs that where released by NVIDIA, and select the appropriate BAR size based on the model identified. This option should be used by default by everyone. The known PCI IDs are listed in the `DeviceRegistry.c` source file.

```
            Manually configure BAR size for specific GPUs:
```
This sub-menu is used instead of the automatic option above, and will take precedence over the automatic option. Use if for example you happen to have some new Turing GPUs that is not recognised from the fixed list of PCI IDs, or if you want to experiment with non-Turing GPUs (for developers or for debugging only).
```
                (1). NVIDIA GeForce RTX 2080 Ti         BAR0 at: 0x8200'0000, size: 16 MiB
```
The base address 0 (BAR0) listed here is needed internally by the DXE driver to enable ReBAR on the GPU. Users should not worry about this address, it is listed here for developers to make sure everything still works as intended.

The sub-menu for manual GPU configuration looks like this:
```
        (P): Select the GPU by PCI ID: 10DE:1E07 (default)
        (S): Select the GPU by PCI ID and Subsystem ID: 10DE:1E07, 1462:3715
        (L): Select the GPU by PCI ID, subystem and bus Location: 10DE:1E07, 1462:3715, 41:00.0
```
This choice tells the UEFI DXE driver how to identify a GPU for manual configuration. This only matters if you have multiple GPUs, or if you add a new GPU to your system. Otherwise the default value (`P`) is fine, and will identify the GPU by its PCI ID. If you have multiple GPUs, then:
* option `P` identifies all GPUs of the same model, like all 2070 Super GPUs
* option `S` identifies all GPUs of the same model, made by the same AIB, like all 2070 Super Pulse cards (from Sapphire)
* option `L` identifies a single GPU by the PCI slot it occupies. The GPU model (that is, the PCI ID) and AIB model (that is the PCI Subsystem ID) must also match.

You can manually configure at most 8 different GPU models this way, but note how usual computer systems can barely fit even 4 GPUs anyway.

```
        (C) Clear per-GPU configuration
```
Manual BAR size configuration above takes precedence over automatic configuration. The option here clears any manual per-GPU configuration, so users can revert back to the automatic BAR size configuration.

```
        (K) Skip BAR size configuration during resume from S3 (sleep)
```
For sleep to work as expected, the GPU BAR size needs to be re-configured on resume from sleep as well. For systems that do not support S3 Sleep, or have other issues with it, this re-configuration on resume (for GPU BAR size) can be disabled with this option. Most users should leave this option enabled.

```
        (O) Enable override for BAR size mask for PCI ReBAR capability
```
This option acts as an work-around for some hardware bug, where a GPU supports ReBAR, but does not report it properly on the PCIe bus when enabled. This option tells the system firmware (motherboard UEFI) to ignore the ReBAR capability reported by the GPU on the bus, and assume ReBAR is supported and enabled. There is no Turing GPU reported with this hardware bug, but an AMD one exists. Most users do not need to use this option (for debugging only).
```
        (P) Select target PCI BAR size, for all (supported) PCI devices (for older boards without ReBAR).
```
Use this for motherboard without ReBAR support from the manufacturer. This will attempt enable ReBAR for such a motherboard, for all PCIe devices that also support ReBAR. The option works in the same way as [ReBarUEFI](https://github.com/xCuri0/ReBarUEFI) works

```
        (W) Show DXE driver configuration (for debugging).
```
NvStrapsReBar.exe communicates the needed configuration to the UEFI DXE driver (NvStrapsReBar.ffs), by writing to an EFI non-volatile variable. This option shows the contents of this variable piece by piece, so developers can see if there are any issues in the driver configuration.

```
        (S) Save configuration changes.
```
Write the current configuration changes to the EFI variable. This will efectively configure the UEFI DXE driver, and take effect on the next boot. For systems without ReBAR support from the manufacturer, remember you need to have "Above 4G Decoding" option enabled in UEFI Setup, and CSM disabled.

```
        (Q) Quit
```
Exit NvStrapsReBar.exe


For `v0.2` and `v0.3`, there is no longer any way to hard-code the GPU or the BAR0 address and PCI bridge address. I have to assume people that ask for this just want to be able to follow some video tutorial, even if it is for an old version of NvStrapsReBar.

Such users can still use `v0.1` , but `v0.1` only works for systems with English language selected in Windows, and does not support resume from sleep (so sleep does not work) ... :(

**Reply by UnidentifiedTag** [reply]
Understood ill leave what you have mentioned above to the end of the video especially the part about v0.1 only supporting English, so if users wish to compile it can only be in English 👍🏻 I’ll setup recording the system I intend to do the demonstration with this Tuesday, I could release the video after ? But I’m guessing everyone wants a Quality demonstration with the best voice over I can offer, I can only do that on the system that currently is out of operation thanks to the rtx 4090 fiasco, might be a blessing in disguise if I get a new one, because the 12v socket has been updated with shorter pins for power, which results in the card not functioning if it isn’t fully inserted…

Worst purchase I have ever made… but that’s another video 😉

**Reply by UnidentifiedTag** [reply]
could you review the preview

many thanks

[https://youtu.be/Dii1_t4q9rU](https://youtu.be/Dii1_t4q9rU)

**Reply by UnidentifiedTag** [reply]
Obviously I have a lot more work to do, like adding credit, song, voice over…

I believe I made a mistake but it worked out for me? Baring in mind I restored the original bios before making the video. Wasn’t I meant to set the settings in the bios 1st before enabling it by the executable? wouldn’t this cause a boot up lock up on some systems. so don’t I need to swap them around. Do the bios 1st then the executable? you see what I’m saying or have you implemented some kind of prevention of lock up happening during boot? because before I’m sure this caused a lock up. I wasn’t really thinking about what to do before what in this video. So basically I need to know is the video good in it’s current form?

**Reply by terminatorul** [reply]
I am still going through the video. There is no sound, and the text is difficult to read, despite I chose 1080p60.

The needed settings in UEFI Setup depend whether the motherboard already has ReBAR from the manufacturer. If it does, users should enable ReBAR if it is not the default, before using `NvStrapsReBar.exe`.

If the motherboard firmware does not have ReBAR implemented from the manufacturer, users need to disable CSM and enable "Above 4G decoding" option (or similar) BEFORE they enable ReBAR using `NvStrapsReBar.exe`.

There is no mechanism in NvStrapsReBar to prevent the wrong UEFI settings in this case (I still hope to implement something like this in the next commit, but it is not foolproof anyway). I understand from the instructions in [ReBarUEFI](https://github.com/xCuri0/ReBarUEFI) that you can still get a 1 GiB BAR without 4G decoding option, even 2 GiB BAR sometimes with the right options.

One thing I noticed on video: when you scroll down in UEFI Tool to find the last DXE driver, you never show the end of the list (as if you already know where the last DXE driver is).

**Reply by terminatorul** [reply]
The video looks good otherwise

**Reply by UnidentifiedTag** [reply]
Unfortunately the Quality isn’t as good as I used an external catcher Device, so it isn’t as clear, it’s still legible. So I believe I’m good to use what I captured.

I was hoping you wouldn’t say this

`If the motherboard firmware does not have ReBAR implemented from the manufacturer, users need to disable CSM and enable "Above 4G decoding" option (or similar) BEFORE they enable ReBAR using NvStrapsReBar.exe.`

so I’m either going to need to edit The footage and make it convincing that I did the bios before opening the rebar executable…

I’ll have to add a voice over to correct this : 

`One thing I noticed on video: when you scroll down in UEFI Tool to find the last DXE driver, you never show the end of the list (as if you already know where the last DXE driver is).
`
Just the preview I shared above has a whole days worth of edits already lol never ends

**Reply by UnidentifiedTag** [reply]
still waiting on my gpu, no word on when i will get it which is causing delays to my video

the 40 series has by far been the worst card ive ever brought

**Reply by UnidentifiedTag** [reply]
updated guide can be found at : [Resizable BAR for Turring GTX 1600 / RTX 2000 GPUs](https://youtu.be/fyOfGbz6WiI?si=r7VCUkYmmBDeCrTE)

I have no Main computer, so could not do voice over. sorry its been so long

**Comment by UnidentifiedTag** [comment]
I have located where the uptake of users has come from : 

you were mentioned in LTT’s tech linked : https://youtu.be/wKcTLmPYlTk?si=SF3oDAzP2qihntCW

**Comment by WWEmerson** [comment]
I have the same motherboard MSI Z170A GAMING M7 and a ASUS DUAL GTX1660 6GB EVO OC. I followed the video steps 1 by 1 and the result when i open the NvStrapsReBar is UEFI DXE driver status: Unconfigured (0x0000000000000028)
Target PCI BAR size: 0 / System default. Still not working :/
